lib/using_restricted_array.rb

```diff
@@ -1,61 +1,224 @@
+
 def find_largest(array, length)
-  raise NotImplementedError
+  max = array[0]
+
+  length.times do |index|
+    if array[index] > max
+      max = array[index]
+    end
+  end
+
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+
+# Approach: 
+# Start with min value as value of first index in array
+# During loop, check if current value is less than current min value
+# If less, chanage min to current value; if not, move on
+# Return the min value after loop is complete
+
+# Complexity:
+# Time complexity: O(n)
+# Explanation - double the number of inputs doubles the population 
+#   therefore, double the amount is filtered till nil is found
+# Space complexity: O(1)
+# Explanation - no matter input, there will only be one number returned
+
 def find_smallest(array, length)
-  raise NotImplementedError
+  min = array[0]
+
+  length.times do |index|
+    if array[index] < min
+      min = array[index]
+    end
+  end
+
+  return min
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+
+# Approach: 
+# Set up first_index to be 0
+# Set up last_index to be index of last element in array
+# Loop as long as first_index is less than last_index
+# During loop, save value of first_index to a temp variable
+# Then replace value of first_index with value of last_index
+# Then replace value of last_index with value in temp variable
+# Increase first_index and decrease last_index by 1
+# After loop, return the reversed array
+
+# Complexity:
+# Time complexity: O(n)
+# Explanation - double the number of inputs doubles the population 
+#   therefore, double the amount is filtered till nil is found
+# Space complexity: O(1)
+# Explanation - no matter input, there will only be one number returned
+
 def reverse(array, length)
-  raise NotImplementedError
+  first_index = 0
+  last_index = length - 1
+
+  while first_index < last_index
+    temp = array[first_index]
+    array[first_index] = array[last_index]
+    array[last_index] = temp
+
+    first_index += 1
+    last_index -= 1
+  end
+
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+
+# Approach: 
+# Set up mid search points and values
+# Continue loop until value found or the counter is greater than length
+# During loop, reassign mid points based on whether value to find
+#   is greater or less than mid points
+# return true/false based on if mid value is the value to find
+
+# Complexity:
+# Time complexity: O(log n)
+# Explanation - cuts population in half during each loop
+#   therefore, adding in twice the input will only increase work by 1
+# Space complexity: O(1)
+# Explanation - no matter input, there will only be one number returned
+
 def binary_search(array, length, value_to_find)
```

Comment:
This works, and is similar to a binary search, but doesn't work quite the same.

Consider for an array of length 10.  If we're looking for an element at index 6.

min_point starts at index 5 and is too small so becomes (10 - 5 ) / 2 + 5 = 8
It checks index 8 and it's too big, then mid_point becomes 4 (which is smaller than a number which was already too small).  

This is a very good attempt , but not quite right.

consider

```ruby
low = 0
high = length - 1
while low <= high
  mid_point = (low + high) / 2

  return true if array[mid_point] == value_to_find

  if array[mid_point] > value_to_find
    high = mid_point - 1
  else
    low = mid_point + 1
  end
end

return false